src/main/groovy/lang/ObjectRange.java

```diff
@@ -408,107 +401,86 @@ public void step(int step, Closure closure) {
             }
 
             @Override
             public void remove() {
-                innerIterator.remove();
+                throw new UnsupportedOperationException();
             }
         };
-        return safeIterator;
     }
 
     /**
-     * convenience class to serve in other methods.
-     * It's not thread-safe, and lazily produces the next element only on calls of hasNext() or next()
+     * Iterator with configurable step size
+     * It's not thread-safe, and lazily produces the current element (including the first one)
+     * only on hasNext() or next() calls.
      */
-    private static class StepIterator implements Iterator {
-        // actual step, can be +1 when desired step is -1 and direction is from high to low
-        private final int step;
+    static class LazySteppingIterator implements Iterator {
         private final ObjectRange range;
-        private int index = -1;
-        private Comparable value;
-        private boolean nextFetched = true;
+        private final boolean isAscending;
+
+        private boolean isNextFetched = false;
+        private Object next = null;
 
-        private StepIterator(ObjectRange range, final int desiredStep) {
+        LazySteppingIterator(ObjectRange range, int desiredStep) {
             if (desiredStep == 0 && compareNotEqual(range.getFrom(), range.getTo())) {
                 throw new GroovyRuntimeException("Infinite loop detected due to step size of 0");
             }
+
             this.range = range;
-            if (range.isReverse()) {
-                step = -desiredStep;
+            if (desiredStep < 0) {
+                isAscending = range.isReverse();
             } else {
-                step = desiredStep;
-            }
-            if (step > 0) {
-                value = range.getFrom();
-            } else {
-                value = range.getTo();
+                isAscending = !range.isReverse();
             }
         }
 
         @Override
-        public void remove() {
-            range.remove(index);
+        public boolean hasNext() {
+            fetchNextIfNeeded();
+            return (next != null) && (isAscending ? compareLessThanEqual(next, range.getTo())
+                                                  : compareGreaterThanEqual(next, range.getFrom()));
         }
 
         @Override
         public Object next() {
-            // not thread safe
-            if (!nextFetched) {
-                value = peek();
-                nextFetched = true;
+            if (!hasNext()) {
+                return null;
             }
-            nextFetched = false;
-            index++;
-            return value;
-        }
 
-        @Override
-        public boolean hasNext() {
-            // not thread safe
-            if (!nextFetched) {
-                value = peek();
-                nextFetched = true;
-            }
-            return value != null;
+            fetchNextIfNeeded();
```

Comment:
@tkruse, thanks for your help, made it lazy again - including the fetching of the first element :)